content/blog/2019-10-23-divergence-optimizations/index.md

```diff
@@ -0,0 +1,231 @@
+
+|     Condition    | Description |
+| ------------- | ------------------- |
+| Instruction is `s2v` | A different scalar value is loaded into each index of a vector register |
+| Instruction is `vload`  | Unknown values are loaded into each element of a vector register |
+| Any data dependency is divergent  | An incoming edge in the dataflow graph is already divergent due to one of the previous conditions |
+
+It's possible that during runtime some vectors marked as divergent might turn out to be convergent. For example, a `vload` may load in contiguous elements with the same values. However, there is no way to optimize for this case statically.
+
+## Divergence Optimizations
+
+### Instruction Swapping
+
+Once we know which instructions are divergent and which are not, we can optimize the code on an instruction-by-instruction basis. In the previous Bril example, every vector instruction is convergent. Therefore we can swap each vector instruction with a more energy-efficient scalar instruction. The optimization is shown below. 
+
+```python
+# initialize vectors -> initialize scalars
+  vec0_s: int = id v0;
+  vec1_s: int = id v1;
+
+# add vectors -> add scalars
+  vec2_s: int = add vec0_s vec1_s;
+
+```
+
+We implement a 'swap table' that matches a vector instruction with a functionally equivalent scalar instruction. An alternate design would be to annotate each original scalar instruction with a vector length and just change the vector length instead of doing a swap. Our swap table is given below along with a description of each instruction reproduced from above.
+
+|     Vector Instruction    | Description | Scalar Instruction |
+| ------------- | ------------- | ------------- |
+| vadd  | c[i] = a[i] + b[i]  | add |
+| vsub  | c[i] = a[i] - b[i]  | sub |
+| vmul  | c[i] = a[i] * b[i]  | mul |
+| vdiv  | c[i] = a[i] / b[i]  | div |
+| s2v  | c = (v0,v1,v2,v3)  | id |
+| s2vb  | c = (v0,v0,v0,v0)  | id |
+| v2s  | c = v[i]  | id |
+| gather  | c[0,1,2,3] = mem[a0,a1,a2,a3]  | lw |
+| scatter  | mem[a0,a1,a2,a3] = a[0,1,2,3]  | sw |
+| vload  | c[i] = mem[i]  | Can't optimize |
+| vstore  | mem[i] = a[i]  | Can't optimize |
+| vcmp  | pred = a[i] == b[i]  | Can't optimize |
+| vphi  | c[i] = pred ? a[i] : b[i]  | Can't optimize |
+
+Notably, we can't optimize across `vload` and `vstore` instructions because different memory addresses are always accessed. However, in certain cases a `gather` and `scatter` can access the exact same memory location if the address vector is the same for each. In this case, the access will be redundant, which will waste memory energy and potentially execution time. Even though we perform the `scatter`/`gather` optimization in the compiler, it is likely that the hardware implementation would also detect this case and avoid the redundant accesses.
```

Comment:
Sounds plausible enough—but are you saying that the hardware is more likely to optimize this than, say, a single-lane vector add? It's possible, I suppose, because it could be worth the added critical-path latency. But it doesn't seem fundamentally different from the other operations you optimize.